File: Proxy/Proxy.go
package main

import "fmt"

type Image interface {
	display()
}

type RealImage struct {
	fileName string
}

func(r *RealImage)loadFromDisk(fileName string){
	fmt.Println("load disk file:",fileName)
}

func(r *RealImage)init(fileName string){
	r.fileName=fileName
	r.loadFromDisk(fileName)
}

func(r *RealImage)display(){
	fmt.Println("displaying:",r.fileName)
}

type ProxyImage struct {
	realImage *RealImage
	fileName string
}

func(p *ProxyImage)display(){
	if(p.realImage==nil){
		p.realImage=new(RealImage)
		p.realImage.init(p.fileName)
	}
	p.realImage.display()
}

func main(){
	proxyImage:=&ProxyImage{fileName:"abcx.jpg"}
	proxyImage.display()
	proxyImage.display()
}

File: Proxy/Proxy.php
<?php
/**
 * Created by PhpStorm.
 * User: jamesyan
 * Date: 2017/3/25
 * Time: 6:46
 */

interface Image{
    public function display();
}


class RealImage implements Image {
    private $fileName;
    public function __construct($fileName)
    {
        $this->fileName=$fileName;
        $this->loadFromDisk($fileName);
    }

    public function display(){
        echo "displaying :".$this->fileName."<br/>";
    }
    public function loadFromDisk($fileName){
        echo "Loding disk".$fileName."<br/>";
    }
}

class ProxyImage implements Image{
    private $realImage,$fileName;
    public function __construct($fileName)
    {
        $this->fileName = $fileName;
    }

    public function display()
    {
        // TODO: Implement display() method.
        if(!$this->realImage instanceof RealImage){
            $this->realImage=new RealImage($this->fileName);
        }
        $this->realImage->display();
    }
}


//demo
//图像从磁盘加载
$image=new ProxyImage("abcx.jpg");
$image->display();
//图像将无法从磁盘加载
$image->display();


